### PR TITLE
tests to check metrics labels

### LINKED ISSUE #95636: More comprehensive test for mapping url paths to metrics labels
In https://github.com/kubernetes/kubernetes/issues/95556 we discovered a bug where kube-apiserver metrics were being incorrectly reported (GET vs LIST problem).
In https://github.com/kubernetes/kubernetes/pull/95562 this was addressed together with test exercising that proper metrics are exposed.

However, this test isn't ideal because if params wouldn't be swapped, it wouldn't catch them.
We should ideally develop a lower-level unit test for mapping url paths to exact values of labels/parameters.

@logicalhan 

/sig apimachinery

### PR DESCRIPTION
**Pull request by leilajal**
**What type of PR is this?**
/ kind bug

**What this PR does / why we need it**:
Adding more comprehensive test for mapping url paths to metrics labels 
**Which issue(s) this PR fixes**:

Fixes #95636
```
fix labels in reporting for kube-apiserver metrics 
```
**Special notes for your reviewer**:
Tests get a baseline of the current metrics, then using test tables to perform various pod and namespace operations. After the operation completes, we get the new generated metric, and compare it to a string representation of the desired metric in the test case.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by leilajal** [comment]
cc @logicalhan

File: test/integration/metrics/metrics_test.go [hunk 1]
```diff
@@ -241,3 +241,255 @@ func TestApiserverMetricsLabels(t *testing.T) {
...
+	makeNamespace := func(labelValue string) *v1.Namespace {
+		return &v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "foo",
+				Labels: map[string]string{"foo": labelValue},
+			},
+		}
+	}
+
+	_, s, closeFn := framework.RunAMaster(framework.NewMasterConfig())
```
Comment (qiutongs) [inline test/integration/metrics/metrics_test.go:375]: Can you rename s to server?

File: test/integration/metrics/metrics_test.go [hunk 2]
```diff
@@ -241,3 +241,255 @@ func TestApiserverMetricsLabels(t *testing.T) {
...
+
+			updatedSamples, err := getSamples(s)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			newSamples := diffMetrics(updatedSamples, baseSamples)
+			found := false
+
+			for _, s := range newSamples {
```
Comment (qiutongs) [inline test/integration/metrics/metrics_test.go:451]: Can you rename the s to sample?

File: test/integration/metrics/metrics_test.go [hunk 3]
```diff
@@ -241,3 +241,255 @@ func TestApiserverMetricsLabels(t *testing.T) {
...
+				if s.String() == tc.want {
+					found = true;
+					break
+				}
+			}
+
+			if !found {
+				t.Fatalf("could not find metric for API call >%s<", tc.name)
```
Comment (qiutongs) [inline test/integration/metrics/metrics_test.go:459]: I suggest to also log "newSamples".

File: test/integration/metrics/metrics_test.go [hunk 4]
```diff
@@ -241,3 +241,255 @@ func TestApiserverMetricsLabels(t *testing.T) {
...
+			},
+		}
+	}
+
+	_, s, closeFn := framework.RunAMaster(framework.NewMasterConfig())
+	defer closeFn()
+
+	client, err := clientset.NewForConfig(&restclient.Config{Host: s.URL, QPS: -1})
+	if err != nil {
+		t.Fatalf("apiserver_request_total metric not exposed")
```
Comment (oomichi) [inline test/integration/metrics/metrics_test.go:274]: Here should be
```
t.Fatalf("Error in create clientset: %v", err)
```
as the same as the existing test TestApiserverMetricsLabels.

File: test/integration/metrics/metrics_test.go [hunk 5]
```diff
@@ -241,3 +241,255 @@ func TestApiserverMetricsLabels(t *testing.T) {
...
```
Comment (oomichi) [inline test/integration/metrics/metrics_test.go:380]: ditto: Here should be
```
t.Fatalf("Error in create clientset: %v", err)
```

**Review by oomichi — COMMENTED** [review]
The additional tests seem good for me.
Just comments about error messages, that seems copy&paste mistakes.

/cc @oomichi

File: test/integration/metrics/metrics_test.go [hunk 6]
```diff
@@ -241,3 +241,255 @@ func TestApiserverMetricsLabels(t *testing.T) {
...
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			newSamples := diffMetrics(updatedSamples, baseSamples)
+			found := false
+
+			for _, s := range newSamples {
+				if s.String() == tc.want {
+					found = true;
```
Comment (oomichi) [inline test/integration/metrics/metrics_test.go:347]: To pass pull-kubernetes-verify job, the following change seems necessary:
```
diff -u ./test/integration/metrics/metrics_test.go.orig ./test/integration/metrics/metrics_test.go
--- ./test/integration/metrics/metrics_test.go.orig	2020-11-03 17:18:04.276518016 +0000
+++ ./test/integration/metrics/metrics_test.go	2020-11-03 17:18:04.276518016 +0000
@@ -344,7 +344,7 @@
 
 			for _, s := range newSamples {
 				if s.String() == tc.want {
-					found = true;
+					found = true
 					break
 				}
 			}
@@ -450,7 +450,7 @@
 
 			for _, s := range newSamples {
 				if s.String() == tc.want {
-					found = true;
+					found = true
 					break
 				}
 			}

Run ./hack/update-gofmt.sh
```

File: test/integration/metrics/metrics_test.go (hunk 5 above)
Comment (leilajal) [reply test/integration/metrics/metrics_test.go]: Thanks, changed.

File: test/integration/metrics/metrics_test.go (hunk 6 above)
Comment (leilajal) [reply test/integration/metrics/metrics_test.go]: Thanks! added the code.

File: test/integration/metrics/metrics_test.go (hunk 4 above)
Comment (leilajal) [reply test/integration/metrics/metrics_test.go]: fixed! thanks

**Comment by oomichi** [comment]
Thanks for updating.

/lgtm

**Comment by leilajal** [comment]
@oomichi @dims @qiutongs what do we need for this? Thanks!

**Pull request merged**